### PR TITLE
Make upsert compaction task more robust to crc mismatch

### PR DESCRIPTION
**Pull request by tibrewalpratik17**
Pushing multiple fixes / enhancements in this patch together for faster reviews.

**Enhancement 1**

Solves Scenario 1 of https://github.com/apache/pinot/issues/13491 where Segment ZK Metadata CRC = Segment CRC deepstore != ValidDocID Bitmap CRC was happening.

The change here is to iterate through all peer servers and see if we can find atleast 1 host with matching CRCs and proceed with the task execution. Assumption is that we have one host which was the leader in updating ZK and uploading to deepstore as well.

We also fail the task-execution if we don't find any such host instead of silently skipping and succeeding the task which we were doing previously.

I tried this fix out for one of our tables and the tasks which were blocked initially due to this issue, it successfully executed compaction after this change and we saw a drop in the row-count and size.
Screenshot 2024-06-27 at 4 20 57 PM

**Enhancement 2**

Solves https://github.com/apache/pinot/issues/13492. The controller fetches validDocIDBitmap from one of the replica servers and compares the CRC in the response against the segment ZK metadata CRC. If they don't match, the segment is not considered for compaction.

The fix is similar to the previous solution. We already retrieve all the bitmaps from all servers. Instead of randomly considering just one, we now iterate through all the bitmaps for each segment. If any of them match the ZK CRC, the segment is then considered for compaction.

I deployed this fix for one of our tables, which unblocked many segments for compaction. Attached is a screenshot showing the drop in rows.
Screenshot 2024-06-29 at 1 00 00 AM

### CODE REVIEW

File: pinot-controller/src/main/java/org/apache/pinot/controller/util/ServerSegmentMetadataReader.java [hunk 1]
```diff
@@ -265,8 +284,12 @@ public List<ValidDocIdsMetadataInfo> getValidDocIdsMetadataFromServer(String tab
...
             JsonUtils.stringToObject(validDocIdsMetadataList, new TypeReference<ArrayList<ValidDocIdsMetadataInfo>>() {
             });
-        for (ValidDocIdsMetadataInfo validDocIdsMetadataInfo : validDocIdsMetadataInfoList) {
-          validDocIdsMetadataInfos.put(validDocIdsMetadataInfo.getSegmentName(), validDocIdsMetadataInfo);
+        for (ValidDocIdsMetadataInfo validDocIdsMetadataInfo: validDocIdsMetadataInfoList) {
+          String segmentName = validDocIdsMetadataInfo.getSegmentName();
+          List<ValidDocIdsMetadataInfo> presentValidDocIdsMetadataInfo =
+              validDocIdsMetadataInfos.computeIfAbsent(segmentName, k -> new ArrayList<>());
+          presentValidDocIdsMetadataInfo.add(validDocIdsMetadataInfo);
+          validDocIdsMetadataInfos.put(segmentName, presentValidDocIdsMetadataInfo);
```
Comment (klsince) [inline pinot-controller/src/main/java/org/apache/pinot/controller/util/ServerSegmentMetadataReader.java:292]: L289-L292 can be inlined as `map.computeIfAbsent().add()`

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompaction/UpsertCompactionTaskExecutor.java [hunk 2]
```diff
@@ -61,30 +66,66 @@ protected SegmentConversionResult convert(PinotTaskConfig pinotTaskConfig, File
...
-    if (!originalSegmentCrcFromTaskGenerator.equals(crcFromDeepStorageSegment)
-        || !originalSegmentCrcFromTaskGenerator.equals(crcFromValidDocIdsBitmap)) {
-      LOGGER.warn("CRC mismatch for segment: {}, expected: {}, actual crc from server: {}", segmentName,
-          crcFromDeepStorageSegment, validDocIdsBitmapResponse.getSegmentCrc());
-      return new SegmentConversionResult.Builder().setTableNameWithType(tableNameWithType).setSegmentName(segmentName)
-          .build();
-    }
-
-    RoaringBitmap validDocIds = RoaringBitmapUtils.deserialize(validDocIdsBitmapResponse.getBitmap());
+    RoaringBitmap validDocIds = null;
```
Comment (klsince) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompaction/UpsertCompactionTaskExecutor.java:75]: nit: move those to the MinionTaskUtils.getValidDocIdsBitmap, for easier reuse

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompaction/UpsertCompactionTaskGenerator.java [hunk 3]
```diff
@@ -218,30 +219,33 @@ public static SegmentSelectionResult processValidDocIdsMetadata(Map<String, Stri
...
-      if (segment == null) {
-        LOGGER.warn("Segment {} is not found in the completed segments list, skipping it for compaction", segmentName);
-        continue;
-      }
+    for (String segmentName: validDocIdsMetadataInfoMap.keySet()) {
+      for (ValidDocIdsMetadataInfo validDocIdsMetadata : validDocIdsMetadataInfoMap.get(segmentName)) {
+        long totalInvalidDocs = validDocIdsMetadata.getTotalInvalidDocs();
+
+        // Skip segments if the crc from zk metadata and server does not match. They may be being reloaded.
+        SegmentZKMetadata segment = completedSegmentsMap.get(segmentName);
```
Comment (klsince) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompaction/UpsertCompactionTaskGenerator.java:227]: this check can be out of the inner for-loop, and continue to next segment `if(!completedSegmentsMap.contains(segmentName))`

File: pinot-controller/src/main/java/org/apache/pinot/controller/util/ServerSegmentMetadataReader.java [hunk 4]
```diff
@@ -265,8 +284,12 @@ public List<ValidDocIdsMetadataInfo> getValidDocIdsMetadataFromServer(String tab
         List<ValidDocIdsMetadataInfo> validDocIdsMetadataInfoList =
             JsonUtils.stringToObject(validDocIdsMetadataList, new TypeReference<ArrayList<ValidDocIdsMetadataInfo>>() {
             });
-        for (ValidDocIdsMetadataInfo validDocIdsMetadataInfo : validDocIdsMetadataInfoList) {
-          validDocIdsMetadataInfos.put(validDocIdsMetadataInfo.getSegmentName(), validDocIdsMetadataInfo);
+        for (ValidDocIdsMetadataInfo validDocIdsMetadataInfo: validDocIdsMetadataInfoList) {
```
Comment (klsince) [inline pinot-controller/src/main/java/org/apache/pinot/controller/util/ServerSegmentMetadataReader.java:287]: code format? as there seems missing spaces

File: pinot-controller/src/main/java/org/apache/pinot/controller/util/ServerSegmentMetadataReader.java (hunk 1 above)
Comment (tibrewalpratik17) [reply pinot-controller/src/main/java/org/apache/pinot/controller/util/ServerSegmentMetadataReader.java]: Done! I thought of using `addAll` as well but then i will have to do if(!empty) then get(0).getSegmentName.
So kept the iterative loop here.

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompaction/UpsertCompactionTaskGenerator.java (hunk 3 above)
Comment (tibrewalpratik17) [reply pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompaction/UpsertCompactionTaskGenerator.java]: Nice catch!

File: pinot-controller/src/main/java/org/apache/pinot/controller/util/ServerSegmentMetadataReader.java (hunk 4 above)
Comment (tibrewalpratik17) [reply pinot-controller/src/main/java/org/apache/pinot/controller/util/ServerSegmentMetadataReader.java]: Ran reformatter on all touched files. You will see some changes going in for that now.

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/MinionTaskUtils.java [hunk 5]
```diff
@@ -206,4 +183,66 @@ public static boolean extractMinionAllowDownloadFromServer(TableConfig tableConf
...
   }
+
+  /**
+   * Returns the validDocID bitmap from the server whose local segment crc matches both crc of ZK metadata and
+   * deepstore copy.
+   */
+  @Nullable
+  public static RoaringBitmap getValidDocIdFromServerMatchingCrc(String tableNameWithType, String segmentName,
+      String validDocIdsType, MinionContext minionContext, String originalSegmentCrcFromTaskGenerator,
+      String crcFromDeepStorageSegment) {
```
Comment (klsince) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/MinionTaskUtils.java:194]: hmm.. should task return early `if (!originalSegmentCrcFromTaskGenerator.equals(crcFromDeepStorageSegment)` 

when these two matched, then we can call this util method to get a validDocIds bitmap from servers that can match the expected crc.

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompaction/UpsertCompactionTaskGenerator.java [hunk 6]
```diff
@@ -60,24 +60,103 @@ public class UpsertCompactionTaskGenerator extends BaseTaskGenerator {
...
+    long invalidRecordsThresholdCount = Long.parseLong(
+        taskConfigs.getOrDefault(UpsertCompactionTask.INVALID_RECORDS_THRESHOLD_COUNT,
+            String.valueOf(DEFAULT_INVALID_RECORDS_THRESHOLD_COUNT)));
+    List<Pair<SegmentZKMetadata, Long>> segmentsForCompaction = new ArrayList<>();
+    List<String> segmentsForDeletion = new ArrayList<>();
+    for (String segmentName : validDocIdsMetadataInfoMap.keySet()) {
+      // check if segment is part of completed segments
+      if (!completedSegmentsMap.containsKey(segmentName)) {
+        LOGGER.warn("Segment {} is not found in the completed segments list, skipping it for compaction", segmentName);
+        break;
```
Comment (klsince) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompaction/UpsertCompactionTaskGenerator.java:79]: should this be `continue`?

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/MinionTaskUtils.java (hunk 5 above)
Comment (tibrewalpratik17) [reply pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/MinionTaskUtils.java:194]: Hmm yeah good point! Though we should fix scenarios where SegmentZKMetadata and deepstore-copy-CRC are not same as a separate issue.

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompaction/UpsertCompactionTaskExecutor.java [hunk 7]
```diff
@@ -60,31 +79,22 @@ protected SegmentConversionResult convert(PinotTaskConfig pinotTaskConfig, File
...
     String originalSegmentCrcFromTaskGenerator = configs.get(MinionConstants.ORIGINAL_SEGMENT_CRC_KEY);
     String crcFromDeepStorageSegment = segmentMetadata.getCrc();
-    String crcFromValidDocIdsBitmap = validDocIdsBitmapResponse.getSegmentCrc();
-    if (!originalSegmentCrcFromTaskGenerator.equals(crcFromDeepStorageSegment)
-        || !originalSegmentCrcFromTaskGenerator.equals(crcFromValidDocIdsBitmap)) {
-      LOGGER.warn("CRC mismatch for segment: {}, expected: {}, actual crc from server: {}", segmentName,
-          crcFromDeepStorageSegment, validDocIdsBitmapResponse.getSegmentCrc());
-      return new SegmentConversionResult.Builder().setTableNameWithType(tableNameWithType).setSegmentName(segmentName)
-          .build();
+    RoaringBitmap validDocIds = originalSegmentCrcFromTaskGenerator.equals(crcFromDeepStorageSegment)
```
Comment (klsince) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompaction/UpsertCompactionTaskExecutor.java:85]: nit: would suggest to check `originalSegmentCrcFromTaskGenerator.equals(crcFromDeepStorageSegment)` firstly and log a msg if they are different for a bit more clarify.

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/MinionTaskUtils.java [hunk 8]
```diff
@@ -206,4 +183,56 @@ public static boolean extractMinionAllowDownloadFromServer(TableConfig tableConf
...
+   * Returns the validDocID bitmap from the server whose local segment crc matches both crc of ZK metadata and
+   * deepstore copy (expectedCrc).
+   */
+  @Nullable
+  public static RoaringBitmap getValidDocIdFromServerMatchingCrc(String tableNameWithType, String segmentName,
+      String validDocIdsType, MinionContext minionContext, String expectedCrc) {
+    String clusterName = minionContext.getHelixManager().getClusterName();
+    HelixAdmin helixAdmin = minionContext.getHelixManager().getClusterManagmentTool();
+    RoaringBitmap validDocIds = null;
+    List<String> servers = MinionTaskUtils.getServers(segmentName, tableNameWithType, helixAdmin, clusterName);
```
Comment (klsince) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/MinionTaskUtils.java:197]: nit: can remove MinionTaskUtils. prefix, as in the same class

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/MinionTaskUtils.java [hunk 9]
```diff
@@ -206,4 +183,55 @@ public static boolean extractMinionAllowDownloadFromServer(TableConfig tableConf
...
+      if (!expectedCrc.equals(crcFromValidDocIdsBitmap)) {
+        // In this scenario, we are hitting the other replica of the segment which did not commit to ZK or deepstore.
+        // We will skip processing this bitmap to query other server to confirm if there is a valid matching CRC.
+        String message = String.format("CRC mismatch for segment: %s, expected value based on task generator: %s, "
+                + "actual crc from validDocIdsBitmapResponse from endpoint %s: %s", segmentName, expectedCrc, endpoint,
+            crcFromValidDocIdsBitmap);
+        LOGGER.warn(message);
+        continue;
+      }
+      validDocIds = RoaringBitmapUtils.deserialize(validDocIdsBitmapResponse.getBitmap());
```
[the remark this answers is not in the document]
Comment (klsince) [reply pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/MinionTaskUtils.java:233]: bump ^ 

looks like this comment was missed?

**Review by klsince — APPROVED** [review]

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/MinionTaskUtils.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (tibrewalpratik17) [reply pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/MinionTaskUtils.java:233]: I have added a `break` right after this which achieves the same.

**Pull request merged**